ibc/docs/ics02.md

```diff
@@ -0,0 +1,156 @@
+	verifyMembershipInnerPayload struct {
+		Height           modules.Height            `json:"height"`
+		DelayTimePeriod  uint64                    `json:"delay_time_period"`
+		DelayBlockPeriod uint64                    `json:"delay_block_period"`
+		Proof            []byte                    `json:"proof"`
+		Path             core_types.CommitmentPath `json:"path"`
+		Value            []byte                    `json:"value"`
+	}
+	verifyMembershipPayload struct {
+		VerifyMembership verifyMembershipInnerPayload `json:"verify_membership"`
+	}
+)
+```
+
+By utilising this pattern of JSON payloads the WASM client itself is able to unmarshal the opaque payload into their own internal protobuf definitions for the implementation of the `ClientState` for example. This allows them to have a much simpler implementation and focus solely on the logic around verification and utilising simple storage.
+
+See: [Types](#types) for more information on the interfaces and types used in the ICS-02 implementation
+
+#### Client Queries
+
+[ICS-24](./ics24.md) instructs that a host must allow for the introspection of both its own `ConsensusState` and `ClientState`. This is done through the `ClientManager`'s `GetHostConsensusState` and `GetHostClientState` methods. These are then used by relayers to:
+
+1. Provide light clients running on counterparty chains the `ConsensusState` and `ClientState` objects they need.
+2. Verify the state of a light client running on a counterparty chain, against the host chain's current `ClientState`
+
+The other queries used by the `ClientManager` involve querying the [ICS-24](./ics24.md) stores to retrieve the `ClientState` and `ConsensusState` stored objects on a per-client basis.
+
+See [Provable Stores](#provable-stores) for more information on how the `ProvableStore`s are used in ICS-02.
+
+## Types
+
+The [ICS-02 specification][ics02] defines the need for numerous interfaces:
+
+1. `ClientState`
+   - `ClientState` is an opaque data structure defined by a client type. It may keep arbitrary internal state to track verified roots and past misbehaviours.
+2. `ConsensusState`
+   - `ConsensusState` is an opaque data structure defined by a client type, used by the
+     validity predicate to verify new commits & state roots. Likely the structure will contain the last commit produced by the consensus process, including signatures and validator set metadata.
+3. `ClientMessage`
+   - `ClientMessage` is an opaque data structure defined by a client type which provides information to update the client. `ClientMessage`s can be submitted to an associated client to add new `ConsensusState`(s) and/or update the `ClientState`. They likely contain a height, a proof, a commitment root, and possibly updates to the validity predicate.
+4. `Height`
+   - `Height` is an interface that defines the methods required by a clients implementation of their own height object `Height`s usually have two components: revision number and revision height.
```

Comment:
Maybe add these to the definitions seciton above?

It's good context as one starts reading everything.

```suggestion
1. `ClientState` - an opaque data structure defined by a client type. It may keep an arbitrary internal state to track verified roots and past misbehaviours.
2. `ConsensusState` - an opaque data structure defined by a client type, used by the validity predicate to verify new commits & state roots. Likely the structure will contain the last commit produced by the consensus process, including signatures and validator set metadata.
3. `ClientMessage` - an opaque data structure defined by a client type which provides information to update the client. `ClientMessage`s can be submitted to an associated client to add new `ConsensusState`(s) and/or update the `ClientState`. They likely contain a height, a proof, a commitment root, and possibly updates to the validity predicate.
4. `Height` - an interface that defines the methods required by a clients implementation of their own height object `Height`s usually have two components: revision number and revision height.
```